lib/cog/pipeline.ex

```diff
@@ -0,0 +1,337 @@
+  end
+
+  @doc """
+  Tells a prepared pipeline to begin executing
+  """
+  def run(pipeline) do
+    GenServer.call(pipeline, :run, :infinity)
+  end
+
+  @doc """
+  Notifies pipeline when processing is finished
+  """
+  def notify(pipeline) do
+    GenServer.cast(pipeline, :teardown)
+  end
+
+  def init([opts]) do
+    try do
+      policy = Keyword.get(opts, :output_policy, @default_output_policy)
+      request = Keyword.fetch!(opts, :request) |> sanitize_request
+      owner = if policy in [:owner, :adapter_owner] do
+        Keyword.fetch!(opts, :owner)
+      else
+        nil
+      end
+      case ConnectionSup.connect() do
+        {:ok, conn} ->
+          {:ok, %__MODULE__{policy: policy, owner: owner,
+                            request: request, conn: conn,
+                            token: Tokens.new(), status: :running}}
+        error ->
+          Logger.error("Failed to connect pipeline #{request.id} to message bus: #{inspect error}")
+          {:stop, error}
+      end
+    rescue
+      e in KeyError ->
+        {:stop, {:missing_option, e.key}}
+    end
+  end
+
+  def handle_call(:run, _from, state) do
+    state = %{state | started: DateTime.utc_now()}
+    case fetch_user(state) do
+      {:ok, user} ->
+        with {:ok, parsed, destinations} <- parse(user, state),
+             {:ok, perms} <- PermissionsCache.fetch(user),
+               user_json <- EctoJson.render(user) do
+          start_pipeline(parsed, destinations, user_json, perms, state)
+        else
+          {:error, {:parse_error, message}} -> start_error_pipeline({:error, :parse_error, message}, user, state)
+          error -> start_error_pipeline(error, user, state)
+        end
+      {:error, :user_not_found}=error ->
+        start_error_pipeline("unregistered-user", error, nil, state)
+      error ->
+        start_error_pipeline(error, nil, state)
+    end
+  end
+  def handle_call(_msg, _from, state) do
+    {:reply, :ignored, state}
+  end
+
+  def handle_cast(:teardown, state) do
+    duration = DateTime.to_unix(DateTime.utc_now, :milliseconds) - DateTime.to_unix(state.started, :milliseconds)
+    Logger.info("Pipeline #{state.request.id} ran for #{duration} ms")
+    Enum.each(state.stages, &(Process.send(&1, {:pipeline_complete, self()}, [])))
+    {:noreply, %{state | status: :done}}
+  end
+  def handle_cast(_, state) do
+    {:noreply, state}
+  end
+
+  # Close bus connection and exit after last stage exits
+  # This avoids closing the connection while one of the sinks is still using it
+  def handle_info({:DOWN, _, _, stage, _}, %__MODULE__{status: :done, stages: [stage]}=state) do
+    Logger.debug("Pipeline #{state.request.id} terminated")
+    Connection.disconnect(state.conn)
+    {:stop, :normal, %{state | stages: []}}
+  end
+  def handle_info({:DOWN, _, _, stage, _}, %__MODULE__{status: :done, stages: stages}=state) do
+    {:noreply, %{state | stages: List.delete(stages, stage)}}
```

Comment:
Would there ever be a situation where the `stage` isn't a member of `stages`? Doesn't look like that would really break things here, but if it could happen, it may be worth logging at least.